### PR TITLE
migrate extra-pages functionality to single-page application

### LINKED ISSUE #472: vTHL4: Migrate extra pages functionality
There used to be a possibility to include extra pages in the navigation header that are markdown or urls. These make it possible to include customizable help etc. pages without code changes in different hosted variants.

- [ ] migrate old extra pages
- [ ] add configurable path for where to host markdowns

### PR DESCRIPTION
**Pull request by okahilak**
closes #472 together with corresponding changes in rems-deploy

### CODE REVIEW

File: src/clj/rems/api/extra_pages.clj [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            [compojure.api.sweet :refer :all]
+            [rems.api.schema :refer :all]
+            [rems.config :refer [env]]
+            [rems.common-util :refer [index-by]]
+            [ring.util.http-response :refer :all]
+            [schema.core :as s])
+  (:import (java.io FileNotFoundException)))
+
+(s/defschema ExtraPage
+  {s/Keyword s/Str})
```
Comment (opqdonut) [inline src/clj/rems/api/extra_pages.clj]: this is shadowing the ExtraPage schema from rems.api.schema, please remove it

File: src/cljs/rems/extra_pages.cljs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ ::enter-page
+ (fn [{:keys [db]} [_ page-id]]
+   {:db (assoc db ::loading? true)
+    ::fetch-extra-page [page-id]}))
+
+(defn- fetch-extra-page [page-id]
+  (fetch (str "/api/extra-pages/" page-id)
+         {:handler #(rf/dispatch [::fetch-extra-page-result %])}))
+
+(rf/reg-fx ::fetch-extra-page (fn [[page-id]] (fetch-extra-page page-id)))
```
Comment (opqdonut) [inline src/cljs/rems/extra_pages.cljs]: you can inline fetch-extra-page here, see docs/architecture/004-reframe-style.md

File: src/cljs/rems/spa.cljs [hunk 3]
```diff
@@ -362,6 +354,10 @@
   (rf/dispatch [:rems.administration.create-workflow/enter-page])
   (rf/dispatch [:set-active-page :rems.administration/create-workflow]))
 
+(secretary/defroute "/extra-pages/:page-id" [page-id]
```
Comment (opqdonut) [inline src/cljs/rems/spa.cljs]: we might need to change this later if some user doesn't like the `/extra-pages/` part in the url, but let's worry about it then

File: src/clj/rems/api/extra_pages.clj [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+(defn- get-extra-page [page-id]
+  (let [allowed-ids (index-by [:id] (filter #(not (:url %)) (:extra-pages env)))]
+    (when (contains? allowed-ids page-id)
+      (let [translations (get-in allowed-ids [page-id :translations])]
+        (into
+         {}
+         (for [lang (keys translations)]
```
Comment (opqdonut) [inline src/clj/rems/api/extra_pages.clj]: could write this as
```
(for [[lang {:keys [filename]}] translations]
  ...)

File: src/clj/rems/api/extra_pages.clj [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+           (let [filename (get-in translations [lang :filename])
+                 extra-pages-path (:extra-pages-path env)
+                 _ (assert extra-pages-path ":extra-pages-path undefined in config")
```
Comment (opqdonut) [inline src/clj/rems/api/extra_pages.clj]: I'd move `extra-pages-path` and the assert to the top-level let, but this works too

File: src/clj/rems/api/extra_pages.clj [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+                 file (io/file extra-pages-path filename)]
+             (if (.isFile file)
+               [lang (slurp file)]
+               (throw (FileNotFoundException. (str "the file specified in extra-pages does not exist: " file)))))))))))
+
+(def extra-pages-api
+  (context "/extra-pages" []
+    :tags ["extra-pages"]
+
+    (GET "/:page-id" []
```
Comment (opqdonut) [inline src/clj/rems/api/extra_pages.clj]: I'd like to see a simple test for this API. One happy case and then the 404 cases (page id unknown, page id known but language unknown, etc.) You can check `rems.api.test-users` for an example. You can override the config in tests using `rems/test-config.edn` or `with-redefs`

File: src/clj/rems/api/extra_pages.clj (hunk 1 above)
Comment (okahilak) [reply src/clj/rems/api/extra_pages.clj]: Done

File: src/cljs/rems/extra_pages.cljs (hunk 2 above)
Comment (okahilak) [reply src/cljs/rems/extra_pages.cljs]: Done, thanks for pointing to the document, I removed the extra layer of indirection as per the document's advice

File: src/cljs/rems/spa.cljs (hunk 3 above)
Comment (okahilak) [reply src/cljs/rems/spa.cljs]: Ok, I didn't do anything about it now.

File: src/clj/rems/api/extra_pages.clj (hunk 4 above)
Comment (okahilak) [reply src/clj/rems/api/extra_pages.clj]: Done

**Pull request merged**